fv3gfs/util/message.py

```diff
@@ -0,0 +1,766 @@
+from dataclasses import dataclass
+from .buffer import Buffer
+from .quantity import Quantity
+from .types import NumpyModule
+from typing import List, Tuple, Optional, Dict
+from enum import Enum
+from .rotate import rotate_scalar_data, rotate_vector_data
+from .utils import device_synchronize, flatten
+import numpy as np
+from uuid import UUID, uuid1
+from .cuda_kernels import (
+    pack_scalar_f64_kernel,
+    pack_vector_f64_kernel,
+    unpack_scalar_f64_kernel,
+    unpack_vector_f64_kernel,
+)
+
+try:
+    import cupy as cp
+except ImportError:
+    cp = None
+
+_CODE_PATH_DEVICE_WIDE_SYNC = False
+_CODE_PATH_CUPY = False
+
+# Keyed cached - key is a str at the moment to go around the fact that
+# a slice is not hashable
+# getting a string from Tuple(slices, rotation, shape, strides, itemsize)
+# IndicesKey = str(Tuple[Any, int, Tuple[int], Tuple[int], int]) #noqa
+INDICES_GPU_CACHE: Dict[str, "cp.ndarray"] = {}
+
+# Simple pool of streams
+STREAM_POOL: List["cp.cuda.Stream"] = []
+
+
+def _pop_stream() -> "cp.cuda.Stream":
+    if len(STREAM_POOL) == 0:
+        return cp.cuda.Stream(non_blocking=True)
+    return STREAM_POOL.pop()
+
+
+def _push_stream(stream: "cp.cuda.Stream"):
+    STREAM_POOL.append(stream)
+
+
+def _slices_length(slices: Tuple[slice]) -> int:
+    """Compute linear size from slices"""
+    length = 1
+    for s in slices:
+        assert s.step is None
+        length *= abs(s.start - s.stop)
+    return length
+
+
+@dataclass
+class MessageMetadata:
+    """Description of a single message."""
```

Comment:
In one or more of these classes it would be good to state what a "message" means, even by saying "MPI message" or something similar. Alternatively, "message" might be too generic of a name - is this some kind of Assignment? e.g. a MessageAssignment or InterRankAssignment or MPIAssignment or otherwise? I didn't expect from the name to see that this has array indexing data in its attributes.

Reply:
Well Message is the technical networking term for a single buffer send through as used in the win32 socket layer, Unity old network system (which was so much better than Unet), etc. For the user of the class, the fact that we have internal indexing, which is only in GPU, is an implementation detail. The metadata is indeed an abuse of linguo because I do not send it with the message...

It's not quite an assignment since it doesn't take care of the networking. It's packing/unpacking mechanism if you want to move away from Message. I'll propose something along those lines.